### DISCUSSION TITLE
Customize x-axis to show a specific number of ticks spaced by a specific number of days

### DISCUSSION THREAD

**Discussion by Praveen-Android** [patrykandpatrick/vico#903] [Questions]
### Question

I am building a graph with formatted dates on the horizontal/x-axis and the x-axis ticks would be spaced out by a day/week/month depending on the time frame selected. I would also like to show only 4 values on x-axis at any point of time and others would be available when scrolled. 

I used `xStep` to get the x-axis ticks spaced by for example a week as shown in the image. This is with `Zoom.Content`

```
        CartesianChartHost(
            chart = rememberCartesianChart(
                  getXStep = { 7f.toDouble() }
            )
       )
```

<img width="231" alt="Screenshot 2024-10-01 at 5 49 25 PM" src="https://github.com/user-attachments/assets/fbe75e2c-72cd-4a65-b930-6074db680912">

And then I am using `rememberVicoZoomState` to just show the first 4 values on x-axis using this:
```
rememberVicoZoomState(
        zoomEnabled = false,
        minZoom = Zoom.x(4f.toDouble())
    )
```

When i do that, I see a crash:
```
FATAL EXCEPTION: main
                            java.lang.IllegalArgumentException: Cannot coerce value to an empty range: 11.708333..10.0.
                            	at kotlin.ranges.RangesKt___RangesKt.coerceIn(_Ranges.kt:1471)
                            	at com.patrykandpatrick.vico.compose.cartesian.VicoZoomState.setValue(VicoZoomState.kt:49)
                            	at com.patrykandpatrick.vico.compose.cartesian.VicoZoomState.setValueRange(VicoZoomState.kt:58)
                            	at com.patrykandpatrick.vico.compose.cartesian.VicoZoomState.update$compose_release(VicoZoomState.kt:106)
                            	at com.patrykandpatrick.vico.compose.cartesian.CartesianChartHostKt.CartesianChartHostImpl$lambda$19(CartesianChartHost.kt:209)
                            	at com.patrykandpatrick.vico.compose.cartesian.CartesianChartHostKt.$r8$lambda$m5Y-f9DthyfY_eq-VXRlZLSxivk(Unknown Source:0)
                            	at com.patrykandpatrick.vico.compose.cartesian.CartesianChartHostKt$$ExternalSyntheticLambda5.invoke(D8$$SyntheticClass:0)
                            	at androidx.compose.ui.draw.DrawBackgroundModifier.draw(DrawModifier.kt:116)
				……..
```

**Note:** The crash only happens when I have the x-axis ticks spaced apart like here (by 7 days) using the step function and then zoom in for 4 values. It doesn't crash when i use the same zoom operation but have the default step (1 day). 

Question: Is there something that I am not doing right in terms of using a combination of step and zoom. 
I basically would like to show a specific count of ticks spaced by a specific number of days on x-axis.

### Vico version(s)

2.0.0-alpha.28

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment]
Hello, @Praveen-Android. Please share a complete, runnable snippet that demonstrates the crash.

**Reply by Praveen-Android** [reply]
Hello @Gowsky Thanks for the reply and your time in looking at this! 
Here is the runnable snippet that leads to the crash:

```

@RequiresApi(Build.VERSION_CODES.O)
val data = mapOf(
  LocalDate.parse("2024-09-04") to 186f,
  LocalDate.parse("2024-09-02") to 186f,
  LocalDate.parse("2024-09-01") to 181f,
  LocalDate.parse("2024-08-28") to 182f,
  LocalDate.parse("2024-08-22") to 188f,
  LocalDate.parse("2024-08-20") to 180f,
  LocalDate.parse("2024-07-25") to 175f
)

@RequiresApi(Build.VERSION_CODES.O)
val formatter: DateTimeFormatter = DateTimeFormatter.ofPattern("MMM d")

@RequiresApi(Build.VERSION_CODES.O)
@Composable
fun Chart1(modifier: Modifier) {
  val modelProducer = remember { CartesianChartModelProducer() }
  val xToDateMapKey = remember { ExtraStore.Key<Map<Float, LocalDate>>() }

  LaunchedEffect(Unit) {
    withContext(Dispatchers.Default) {
      val xToDates = data.keys.associateBy { it.toEpochDay().toFloat() }
      modelProducer.runTransaction {
        lineSeries { series(xToDates.keys, data.values) }
        extras { it[xToDateMapKey] = xToDates }
      }
    }
  }
}

@RequiresApi(Build.VERSION_CODES.O)
@Composable
private fun ComposeChart1(
  modelProducer: CartesianChartModelProducer,
  modifier: Modifier,
  xToDateMapKey: ExtraStore.Key<Map<Float, LocalDate>>
) {
  val marker = rememberMarker()

  // I want the scroll be disabled. I will programmatically scroll when clicked on a button
  val scrollState = rememberVicoScrollState(
    scrollEnabled = false,
    initialScroll = Scroll.Absolute.End
  )

  // I want the graph to always show only 4 x-axis ticks at any given time.
  val zoomState = rememberVicoZoomState(
    zoomEnabled = false,
    minZoom = Zoom.x(4.0)
  )

  // I want the x-axis to show ticks in steps of 7 days (for example but could be a month)
  val step = 7.0

  val yMin = data.values.minOrNull()?: 0f
  val yMax = data.values.maxOrNull()?: 0f

  CartesianChartHost(
    chart = rememberCartesianChart(
      rememberLineCartesianLayer(
        LineCartesianLayer.LineProvider.series(
          rememberLine(
            // color to fill the line with
            fill = remember { LineCartesianLayer.LineFill.single(fill(Color.Blue)) },
            // do we want to color fill the area between the line and x-axis line
            areaFill = null,
            // should the lines be curved or straight when traversing through data points
            pointConnector = remember { LineCartesianLayer.PointConnector.cubic(curvature = 0f) }
          ),
        ),
        axisValueOverrider = AxisValueOverrider.fixed(
          minY = floor(yMin.toDouble()),
          maxY = ceil(yMax.toDouble())
        )
      ),
      // y axis configuration
      endAxis = rememberEndAxis(
        line = null,
        tick = null,
        verticalLabelPosition = VerticalAxis.VerticalLabelPosition.Top,
        itemPlacer = remember { VerticalAxis.ItemPlacer.count( { 5 }) },
        valueFormatter = { value: Double, _, _ -> value.toInt().toString() }
      ),
      bottomAxis = rememberBottomAxis(
        line = rememberAxisLineComponent(shape = DashedShape()),
        tick = null,
        guideline = null,
        valueFormatter = { value: Double, chartValues: ChartValues, _ ->
          (chartValues.model.extraStore[xToDateMapKey][value.toFloat()]?: LocalDate.ofEpochDay(value.toLong())).format(formatter)
        }
      ),
      getXStep = { step },
      marker = marker
    ),
    modelProducer = modelProducer,
    modifier = modifier,
    zoomState = zoomState,
    scrollState = scrollState,
    runInitialAnimation = true
  )
}
```

**Crash:**
```
 FATAL EXCEPTION: main
                            Process: com.patrykandpatrick.vico, PID: 14784
                            java.lang.IllegalArgumentException: Cannot coerce value to an empty range: 19.25..10.0.
                            	at kotlin.ranges.RangesKt___RangesKt.coerceIn(_Ranges.kt:1471)
                            	at com.patrykandpatrick.vico.compose.cartesian.VicoZoomState.setValue(VicoZoomState.kt:49)
                            	at com.patrykandpatrick.vico.compose.cartesian.VicoZoomState.setValueRange(VicoZoomState.kt:58)
                            	at com.patrykandpatrick.vico.compose.cartesian.VicoZoomState.update$compose_debug(VicoZoomState.kt:106)
                            	at com.patrykandpatrick.vico.compose.cartesian.CartesianChartHostKt.CartesianChartHostImpl$lambda$19(CartesianChartHost.kt:209)
                            	at com.patrykandpatrick.vico.compose.cartesian.CartesianChartHostKt.$r8$lambda$m5Y-f9DthyfY_eq-VXRlZLSxivk(Unknown Source:0)
                            	at com.patrykandpatrick.vico.compose.cartesian.CartesianChartHostKt$$ExternalSyntheticLambda5.invoke(D8$$SyntheticClass:0)
                            	at androidx.compose.ui.draw.DrawBackgroundModifier.draw(DrawModifier.kt:116)
```

**Reply by Praveen-Android** [reply]
**Update:** I looked a little deeper and realized that I had a different understanding of how `zoom.x` works. What i thought was if I added a xstep of 7, then a `zoom.x(4.0)` would zoom in to 4 ticks each spaced a week apart. But it actually would be `zoom.x(28.0)` based on the `zoom.x` calculation function:
`bounds.width() * (context.chartValues.xStep / x).toFloat() / horizontalDimensions.xSpacing`

So doing a zoom.x(28.0) does zoom down to 4 ticks spaced 7 days apart and no crash:
<img width="240" alt="image" src="https://github.com/user-attachments/assets/ef61e2cf-f1af-4336-893a-c575039340a5">

Here is how I accomplished some things:
- Show dates on x-axis in steps of a week/month
    -`getXStep = { xStep }`
- Show only a specific number of ticks on x-axis
    ```
    rememberVicoZoomState(
        zoomEnabled = false,
        minZoom = Zoom.x( 28.0)
        )
    ) 
    ```
- Ability to programmatically scroll forward/backward by the same number of ticks (ex: 4) that zoom would show
```
    coroutineScope.launch {
        scrollState.animateScroll(Scroll.Relative.x(28.0)) // (and use -28.0 to scroll backward)
        }
```
**Questions**:

- Is Zoom the right way to always show a specific number of ticks on x-axis or is there an better way? 
    - What i observed is that using a combination of zoom and scroll like i mentioned above doesn't always show the exact values or position them properly on the axis. Screenshot here:
    -<img width="247" alt="image" src="https://github.com/user-attachments/assets/021bc193-d753-4ce0-b75d-9e04bb9c1b62">

- Is there a way to Zoom / Scroll to a particular value/date on the x-axis using zoomstate / scrollstate?

Thanks!

**Reply by Gowsky** [reply]
> What i thought was if I added a xstep of 7, then a `zoom.x(4.0)` would zoom in to 4 ticks each spaced a week apart. But it actually would be `zoom.x(28.0)` based on the `zoom.x` calculation function:

Yes, as per the documentation, `Zoom.x` makes a specific number of _x_ units visible. You're looking to show 28 units = 28 days = 4 weeks. The _x_ step doesn't change that.

> Is Zoom the right way to always show a specific number of ticks on x-axis or is there an better way?

You're not trying to show a specific number of ticks but rather to show a specific number of _x_ units in the viewport. This is merely reflected by the axis, which, in its current configuration, displays one label per week. There could just as well be no axis. With that said, yes, `Zoom.x` is the correct API here.

> What i observed is that using a combination of zoom and scroll like i mentioned above doesn't always show the exact values or position them properly on the axis. 

The snippet that you've shared doesn't demonstrate this, so please share one that does. You may be misuing `Scroll`.

> Is there a way to Zoom / Scroll to a particular value/date on the x-axis using zoomstate / scrollstate?

What do you mean by "zoom/scroll"? If you want to scroll to a particular date, you can do so using `Scroll.Absolute.x`. (Note the `Absolute` — this is different from what you've mentioned that you're using.) 

Also, your usage of `AxisValueOverrider` is not entirely correct. Please see [here](https://patrykandpatrick.com/vico/wiki/cartesian-charts/layers/). (This is for version `2.0.0-beta.1`, where `AxisValueOverrider` has been renamed to `CartesianLayerRangeProvider`, but the same principles apply.)
